consensus/reactor.go

```diff
@@ -557,1170 +536,797 @@ OUTER_LOOP:
-						Height:  prs.Height,
-						Round:   commit.Round,
-						Type:    tmproto.PrecommitType,
-						BlockID: commit.BlockID,
-					}))
-					time.Sleep(conR.conS.config.PeerQueryMaj23SleepDuration)
+			if prs.CatchupCommitRound != -1 && prs.Height > 0 && prs.Height <= r.conS.blockStore.Height() &&
+				prs.Height >= r.conS.blockStore.Base() {
+				if commit := r.conS.LoadCommit(prs.Height); commit != nil {
+					r.stateCh.Out() <- p2p.Envelope{
+						To: ps.peerID,
+						Message: &tmcons.VoteSetMaj23{
+							Height:  prs.Height,
+							Round:   commit.Round,
+							Type:    tmproto.PrecommitType,
+							BlockID: commit.BlockID.ToProto(),
+						},
+					}
+
+					time.Sleep(r.conS.config.PeerQueryMaj23SleepDuration)
 				}
 			}
 		}
 
-		time.Sleep(conR.conS.config.PeerQueryMaj23SleepDuration)
-
+		time.Sleep(r.conS.config.PeerQueryMaj23SleepDuration)
 		continue OUTER_LOOP
 	}
 }
 
-func (conR *Reactor) peerStatsRoutine() {
-	for {
-		if !conR.IsRunning() {
-			conR.Logger.Info("Stopping peerStatsRoutine")
-			return
-		}
+func (r *Reactor) processPeerUpdate(peerUpdate p2p.PeerUpdate) {
+	r.Logger.Debug("received peer update", "peer", peerUpdate.PeerID, "status", peerUpdate.Status)
 
-		select {
-		case msg := <-conR.conS.statsMsgQueue:
-			// Get peer
-			peer := conR.Switch.Peers().Get(msg.PeerID)
-			if peer == nil {
-				conR.Logger.Debug("Attempt to update stats for non-existent peer",
-					"peer", msg.PeerID)
-				continue
-			}
-			// Get peer state
-			ps, ok := peer.Get(types.PeerStateKey).(*PeerState)
-			if !ok {
-				panic(fmt.Sprintf("Peer %v has no state", peer))
-			}
-			switch msg.Msg.(type) {
-			case *VoteMessage:
-				if numVotes := ps.RecordVote(); numVotes%votesToContributeToBecomeGoodPeer == 0 {
-					conR.Switch.MarkPeerAsGood(peer)
-				}
-			case *BlockPartMessage:
-				if numParts := ps.RecordBlockPart(); numParts%blocksToContributeToBecomeGoodPeer == 0 {
-					conR.Switch.MarkPeerAsGood(peer)
-				}
-			}
-		case <-conR.conS.Quit():
-			return
+	psKey := string(peerUpdate.PeerID)
 
-		case <-conR.Quit():
+	switch peerUpdate.Status {
+	case p2p.PeerStatusNew, p2p.PeerStatusUp:
+		// Do not allow starting new broadcasting goroutines after reactor shutdown
+		// has been initiated. This can happen after we've manually closed all
+		// peer goroutines and closed r.closeCh, but the router still sends in-flight
+		// peer updates.
+		if !r.IsRunning() {
 			return
 		}
-	}
-}
 
-// String returns a string representation of the Reactor.
-// NOTE: For now, it is just a hard-coded string to avoid accessing unprotected shared variables.
-// TODO: improve!
-func (conR *Reactor) String() string {
-	// better not to access shared variables
-	return "ConsensusReactor" // conR.StringIndented("")
-}
+		// save the peer if the ID is non-empty and we haven't seen this peer before
+		if psKey != "" && !r.peers.Has(psKey) {
+			peerState := NewPeerState(r.Logger, peerUpdate.PeerID)
+			r.peers.Set(psKey, peerState)
```

Comment:
There's a race condition here, since `Has()`, `Get()`, and `Set()` are not called in the same critical section (i.e. while holding a mutex). It's possible for a different goroutine to call `Delete()` on the peer between `Has()` and `Get()`, causing the reactor to panic. There's a few different ways to handle this, depending on our needs, but in this case it's probably sufficient to add e.g. `peers.Ensure()` which atomically returns the existing peer state if any or otherwise creates a new one, stores it in the map, and returns it.

Reply:
`r.peers` is thread-safe, no? So why do we need an additional mutex?

Reply:
Ohh you mean a mutex around the entire block of execution -- not just API calls to `peers`?

Reply:
No wait, this actually isn't true, since `processPeerUpdate` is not called concurrently and it is the only method that mutates the map.

Either way, I'd simplify the code a bit by doing `peerState, ok := Get(); if !ok { peerState = NewPeerState(); Set() }`, which avoids a panic case.